activesupport/lib/active_support/cache/redis_cache_store.rb

```diff
@@ -0,0 +1,389 @@
+# frozen_string_literal: true
+
+begin
+  require "redis"
+  require "redis/distributed"
+rescue LoadError
+  warn "The Redis cache store requires the redis gem. Please add it to your Gemfile."
+  raise
+end
+
+# Prefer the hiredis driver but don't require it.
+begin
+  require "redis/connection/hiredis"
+rescue LoadError
+end
+
+require "digest/md5"
+require "active_support/core_ext/marshal"
+
+module ActiveSupport
+  module Cache
+    # Redis cache store.
+    #
+    # * Supports vanilla Redis, hiredis, and Redis::Distributed.
+    # * Supports Memcached-like sharding across Redises with Redis::Distributed.
+    # * Fault tolerant. If the Redis server is unavailable, no exceptions are
+    #   raised. Cache fetches are all misses and writes are dropped.
+    # * Local cache. Hot in-memory primary cache within block/middleware scope.
+    # * `read_/write_multi` support for Redis mget/mset. Use Redis::Distributed
+    #   4.0.1+ for distributed mget support.
+    # * `delete_matched` support for Redis KEYS globs.
+    class RedisCacheStore < Store
+      # Keys are truncated with their own MD5 digest if they exceed 1kB
+      MAX_KEY_BYTESIZE = 1024
+
+      DEFAULT_REDIS_OPTIONS = {
+        connect_timeout:    20,
+        read_timeout:       2,
+        write_timeout:      2,
+        reconnect_attempts: 0,
+      }
```

Comment:
Conservative defaults. Consider lowering read/write timeouts to 0.5 or 1 second. Typically cheaper to regenerate a cached value than wait more than a second for it.

Reply:
Lowered these to 1 second. Unclear whether most Redis clients can accommodate sub-second timeouts.